be1-go/validation/schema_validator.go

```diff
@@ -0,0 +1,107 @@
+package validation
+
+import (
+	"fmt"
+	"path/filepath"
+	"strings"
+
+	"student20_pop/message"
+
+	"github.com/xeipuuv/gojsonschema"
+	"golang.org/x/xerrors"
+)
+
+type SchemaValidator struct {
+	schemas map[string]*gojsonschema.Schema
+}
+
+type schema struct {
+	name string
+	path string
+}
+
+const (
+	GenericMsgSchema string = "genericMsgSchema"
+	DataSchema       string = "dataSchema"
+)
+
+// RegisterSchema adds a JSON schema specified by the path
+func (s *SchemaValidator) RegisterSchema(schema schema) error {
+	msgLoader := gojsonschema.NewReferenceLoader(schema.path)
+	msgSchema, err := gojsonschema.NewSchema(msgLoader)
+	if err != nil {
+		return xerrors.Errorf("failed to load the json schema with path `%s`: %v", schema.path, err)
+	}
+
+	s.schemas[schema.name] = msgSchema
+	return nil
+}
+
+// VerifyJson verifies that the 'byteMessage' follow the schema protocol of name 'schemaName',
+// it returns an error if it is not the case.
+func (s *SchemaValidator) VerifyJson(byteMessage []byte, schemaName string) error {
+	// Validate the Json "byteMessage" with a schema
+	messageLoader := gojsonschema.NewBytesLoader(byteMessage)
+	resultErrors, err := s.schemas[schemaName].Validate(messageLoader)
+	if err != nil {
+		return &message.Error{
+			Code:        -4,
+			Description: fmt.Sprintf("failed to validate schema: %s", err.Error()),
+		}
+	}
+
+	errorsList := resultErrors.Errors()
+	descriptionErrors := ""
+	// Concatenate all error descriptions
+	for index, e := range errorsList {
+		descriptionErrors += fmt.Sprintf(" (%d) %s", index+1, e.Description())
+	}
+
+	if len(errorsList) > 0 {
+		return &message.Error{
+			Code:        -4,
+			Description: descriptionErrors,
+		}
+	}
+
+	return nil
+}
+
+// NewSchemaValidator returns a Schema Validator
+func NewSchemaValidator() (*SchemaValidator, error) {
+	// Import the Json schemas defined in the protocol section
+	protocolPath, err := filepath.Abs("../protocol")
```

Comment:
This would fail the test if you check for errors returned from `RegisterSchema`.

Reply:
```
=== RUN   TestOrganizer_SchemaLoading
    schema_validator_test.go:12:
                Error Trace:    schema_validator_test.go:12
                Error:          Received unexpected error:
                                failed to register schema:
                                    student20_pop/validation.NewSchemaValidatorWithSchemas
                                        /Users/narula/dev/student20_pop/be1-go/validation/schema_validator.go:106
                                  - failed to load the json schema with path `file:///Users/narula/dev/student20_pop/be1-go/protocol/genericMessage.json`:
                                    student20_pop/validation.(*SchemaValidator).RegisterSchema
                                        /Users/narula/dev/student20_pop/be1-go/validation/schema_validator.go:33
                                  - open /Users/narula/dev/student20_pop/be1-go/protocol/genericMessage.json: no such file or directory
                Test:           TestOrganizer_SchemaLoading
--- FAIL: TestOrganizer_SchemaLoading (0.00s)
```

Reply:
This has another assumption that the executable is always invoked from `be1-go` subdirectory. That's not the case for executing tests and might also not hold when the server is deployed.

I'd suggest accepting the protocol directory as a command line argument instead and passing it as an argument to `NewSchemaValidator` and `NewSchemaValidatorWithSchemas`.

Reply:
I added a command line argument for the protocol path and if no value is passed the server loads the schemas using the URL of the protocol (on the GitHub website)